lib/src/main/java/io/ably/lib/types/DecodingContext.java

```diff
@@ -0,0 +1,38 @@
+package io.ably.lib.types;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.nio.charset.StandardCharsets;
+
+public class DecodingContext {
+
+	private String lastMessageString;
+	private byte[] lastMessageBinary;
+
+	public DecodingContext(Map<String, PluggableCodec> codecs)
+	{
+		lastMessageBinary = null;
+		lastMessageString = null;
+		this.codecs = codecs;
+	}
+
+	public Map<String, PluggableCodec> codecs;
```

Comment:
I personally am not a fan of public fields in languages like Java. In this project, however, we have them all over the place so I am adhering to the existing coding style. Changing it piecemeal would probably cause bigger mess.

I am not really sure I understand why you think that field does not belong to the `DecodingContext` class. Can you elaborate? Where does it belong?

Reply:
I get your point around existing, implicit coding style (i.e. inherited from surrounding code). However, I would rather that new code adheres to accepted Java conventions... in lieu of us producing an explicit coding style and/or adopting a tool like checkstyle.

The `codecs` field is just wide open. It's never actively used by this class and, on top of that, the reference to the map is mutable, as are the contents! Perhaps you could refine to what's really required? e.g. make the field private and then expose a public method `PluggableCodec findCodec(String codecName)` and define that if the codec is not found then `null` will be returned, which is then a decision for the calling code. What do you think?

Reply:
As mentioned I personally prefer the architecture you describe. I still think that having drastically different standards for the new code would result in a bad mixture of coding styles but that decision is yours to make.

Before doing that change, though, let's see whether we are not missing something.  We have a lot of public collections all over the place (e.g. only public `Map`s are a couple of dozens) and maybe there is a reason for that. @paddybyers what would that be?

Reply:
Yeah, @paddybyers and I have had quite a few discussions already around my observations of the Java codebase and where things are, perhaps, a little 'over exposed'. Those interfaces will have to stand as they are until we get the bandwidth to refactor / deprecate / etc.. In the meantime, however, we must strive to make any new interfaces clean and minimal - especially when they form part of our front line public API. That approach has got to prevent some bugs from making their way in to both our code and that of developers using our code.